### ISSUE TITLE
tofu reco with flat-field correction on multiple devices

### ISSUE THREAD

**Issue by xeiwoka** [ufo-kit/tofu#79]
tofu reco (laminographic axis tilt 32.5°) considerably slows down (~2x duration) when using flat-field correction (ffc) on multiple GPUs. Without ffc and using --only-bp the durations increase by only some tens of percent.

This also happens when one uses the device 1 out of 0,1.

Examples on 2x Tesla K20m:
# Using device 1 out of 0,1:
#DEBUG Duration:  --gpus 1 --num-gpu-threads 1
tofu reco --burst 8 --retry-timeout 10 --retries 6 --number 4200 --step 1 --darks /run/shm/sc4593//set15_SCW3_zoneL_mid_325nm_35keV_real_40mm/darkend0000.edf --dark-scale .02000 --flats /run/shm/sc4593//set15_SCW3_zoneL_mid_325nm_35keV_real_40mm/ref*_0000.edf --flats2 /run/shm/sc4593/set15_SCW3_zoneL_mid_325nm_35keV_real_40mm/ref*_4200.edf --projections /run/shm/sc4593/set15_SCW3_zoneL_mid_325nm_35keV_real_40mm/set15_SCW3_zoneL_mid_325nm_35keV_real_40mm*.edf --absorptivity --pixel-size .000000325 --projection-padding-mode clamp_to_edge --genreco-padding-mode clamp_to_edge --x-region=-1280,1280,1 --y-region=-1280,1280,1 --overall-angle -360 --verbose --slices-per-device 125 --axis-angle-y 0 --axis-angle-x 32.5 --output /data/visitor/sc4593/id19/Lukas_8bit/./set15_SCW3_zoneL_mid_325nm_35keV_real_40mm/xySlices --fix-nan-and-inf --z-parameter z --region=-900.000,900.000,14.400 --center-position-x 1316.25 --center-position-z 1080.5 --volume-angle-x 0 --volume-angle-y 0 --gpus 1 --num-gpu-threads 1
DEBUG: UFO duration: 698.37 s (too slow!)
[similar duration times obtained using two GPUs]

# now restricting to device 1:
CUDA_VISIBLE_DEVICES=1
tofu reco --burst 8 --retry-timeout 10 --retries 6 --number 4200 --step 1 --darks /run/shm/sc4593//set15_SCW3_zoneL_mid_325nm_35keV_real_40mm/darkend0000.edf --dark-scale .02000 --flats /run/shm/sc4593//set15_SCW3_zoneL_mid_325nm_35keV_real_40mm/ref*_0000.edf --flats2 /run/shm/sc4593/set15_SCW3_zoneL_mid_325nm_35keV_real_40mm/ref*_4200.edf --projections /run/shm/sc4593/set15_SCW3_zoneL_mid_325nm_35keV_real_40mm/set15_SCW3_zoneL_mid_325nm_35keV_real_40mm*.edf --absorptivity --pixel-size .000000325 --projection-padding-mode clamp_to_edge --genreco-padding-mode clamp_to_edge --x-region=-1280,1280,1 --y-region=-1280,1280,1 --overall-angle -360 --verbose --slices-per-device 125 --axis-angle-y 0 --axis-angle-x 32.5 --output /data/visitor/sc4593/id19/Lukas_8bit/./set15_SCW3_zoneL_mid_325nm_35keV_real_40mm/xySlices --fix-nan-and-inf --z-parameter z --region=-900.000,900.000,14.400 --center-position-x 1316.25 --center-position-z 1080.5 --volume-angle-x 0 --volume-angle-y 0 --gpus 0 --num-gpu-threads 1
DEBUG: UFO duration: 376.32 s (OK!)

**Comment by tfarago** [comment]
Thanks for the commands, I will look into it.

**Comment by milkdrop** [comment]
Hello, 

I can report the similar behavior on our server (2x P5000): 
1:
time tofu tomo --projections tomo --flats flats --darks darks --output sli/sli_%04i.tif --axis 1283 --absorptivity --fix-nan-and-inf
real    0m34.000s
2:
time tofu reco --projections tomo --flats flats --darks darks --output sli/sli --log log.txt --center-position-x 1283 --absorptivity --fix-nan-and-inf --overall-angle 180 --verbose --region 103,-104,-1
real    0m59.583s
3:
time tofu lamino --projections tomo --flats flats --darks darks --output sli/sli --log log.txt --center-position-x 1283 --lamino-angle 90 --absorptivity --fix-nan-and-inf --overall-angle 180 --axis 1283,90
real    1m2.463s

**Comment by milkdrop** [comment]
I'd like to bump this one. I have reasons to believe that ufo-launch has problems on systems with two GPUs (at least on ours 2xP5000). I will do some benchmarking and will create topic in the ufo-filters subforum if/when I get evidences.

**Comment by tfarago** [comment]
Please run the `tofu reco` with `--verbose` flag. It could be that the volume you are reconstructing doesn't fit into the GPU memory of your two cards, i.e. it needs to run a second pass, which of course slows things down.

**Comment by tfarago** [comment]
Please run the `tofu reco` with `--verbose` flag. It could be that the volume you are reconstructing doesn't fit into the GPU memory of your two cards, i.e. it needs to run a second pass, which of course slows things down.

**Comment by tfarago** [comment]
Could you run these commands?
`time tofu tomo --width 2016 --height 3000 --number 1200 --dry-run --axis 1008.5`
and
`time tofu reco --width 2016 --height 1200 --number 3000 --dry-run --region=-1008,1008,1 --center-position-x 1008.5 --center-position-z 1008.5 --verbose --overall-angle 180`
Then post the output of the `reco` command. For me the times are 41 vs 55 seconds in favor of `reco`. I get similar results for real files but first let's make sure there is nothing weird going on outside of the file system.

**Comment by milkdrop** [comment]
Hi Tomas, 

here are the results for 
server IOC (2x P5000 each 16 GB RAM 9 TFLOPs, 256 GB RAM) and server OPI1117-001 (1x RTX4000 8 GB RAM, 7 TFLOPs, 128 GB RAM) 

Test 4:
time tofu tomo --width 2016 --height 3000 --number 1200 --dry-run --axis 1008.5
IOC (2x P5000)
INFO: Execution time: 100.574483 s (real 1m42.277s)
OPI (1x RTX4000)
INFO: Execution time: 106.288033 s (real 1m47.056s)

Test 5:
time tofu reco --width 2016 --height 1200 --number 3000 --dry-run --region=-1008,1008,1 --center-position-x 1008.5 --center-position-z 1008.5 --verbose --overall-angle 180
IOC
DEBUG: UFO duration: 126.13 s
DEBUG: Total duration: 128.51 s
DEBUG: UFO performance: 194.89 GUPS
DEBUG: Total performance: 191.27 GUPS
real 2m10.828s

OPI
DEBUG: UFO duration: 229.80 s
DEBUG: Total duration: 230.02 s
DEBUG: UFO performance: 106.96 GUPS
DEBUG: Total performance: 106.86 GUPS
real 3m50.901s

So the pure "reco" is twice faster. My question is about ufo-launch. Are filters supposed to be faster on multiple GPUs? My other tests:
for input data size - 31 GB raw data and 62 GB 32bit floating point. Local disk to local disk, IOC has hardware RAID0, OPI has a very good SSD:
Test2:
time tofu sinos --projection proj-ffc/ --output sin/sin-%04i.tif  (Local → local)
3 minutes both

Test3:
time ufo-launch read path=sin ! fft dimensions=2 ! filter-stripes sigma=1.0 ! ifft dimensions=2 crop-width=2560 crop-height=3000 ! write filename=sin-filt/sin-%04i.tif
IOC1131:
2160 items processed ...
Finished in 89.08065s (real 1m29.747s)
OPI1117:
2160 items processed ...
Finished in 147.34256s (real 2m27.431s)

Test4:
IOC 
time tofu preprocess --projections proj-ffc/ --output proj-PR/projPR-%04i.tif --energy 18 --propagation-distance 0.2 --pixel-size 6.5e-6 --regularization-rate 2.5 --projection-filter none  
real    5m40.245s
OPI
time tofu preprocess --projections proj-ffc/ --output proj-PR/projPR-%04i.tif --energy 18 --propagation-distance 0.2 --pixel-size 6.5e-6 --regularization-rate 2.5 --projection-filter none
real    3m13.942s

Test5 (sin filter again)
time ufo-launch read path=sin ! fft dimensions=2 ! filter-stripes sigma=1.0 ! ifft dimensions=2 crop-width=2560 crop-height=3000 ! write filename=sin-filt/sin-%04i.tif
IOC:
Finished in 208.27370s     real    3m29.744s
OPI:
Finished in 172.15494s    real    2m52.229s

What I see is that server with less GPU power and in general inferior characteristics is doing ufo-launch filters as fast as server with two GPUs (twice computing power) and hardware RAID0... Any ideas why is that? 

Regards,
Sergei.

**Comment by milkdrop** [comment]
OK, just discussed results with Chris who is pretty good with the hardware and he is not surprised. Ultra-fast SSDs from Dell are faster then old hardware RAID0 and RTX4000 seems to be significantly better then P5000. Hence, the only case when two GPU are faster in all tests is when algorithm requires a lot of RAM. Does it make sense, Tomas?

**Comment by tfarago** [comment]
I am not surprised that "small" operations don't scale, because there the cards just sit and wait for data, but that's not the case of `tofu tomo` and `tofu reco` unless the reconstructed volume is tiny.

**Comment by tfarago** [comment]
But I see that with 2 cards you are almost 2x faster on the P5000 machine, so that's good news.

**Comment by tfarago** [comment]
> tofu reco (laminographic axis tilt 32.5°) considerably slows down (~2x duration) when using flat-field correction (ffc) on multiple GPUs. Without ffc and using --only-bp the durations increase by only some tens of percent.
> 
> This also happens when one uses the device 1 out of 0,1.
> 
> Examples on 2x Tesla K20m:
> # Using device 1 out of 0,1:
> 
> #DEBUG Duration: --gpus 1 --num-gpu-threads 1
> tofu reco --burst 8 --retry-timeout 10 --retries 6 --number 4200 --step 1 --darks /run/shm/sc4593//set15_SCW3_zoneL_mid_325nm_35keV_real_40mm/darkend0000.edf --dark-scale .02000 --flats /run/shm/sc4593//set15_SCW3_zoneL_mid_325nm_35keV_real_40mm/ref*_0000.edf --flats2 /run/shm/sc4593/set15_SCW3_zoneL_mid_325nm_35keV_real_40mm/ref*_4200.edf --projections /run/shm/sc4593/set15_SCW3_zoneL_mid_325nm_35keV_real_40mm/set15_SCW3_zoneL_mid_325nm_35keV_real_40mm*.edf --absorptivity --pixel-size .000000325 --projection-padding-mode clamp_to_edge --genreco-padding-mode clamp_to_edge --x-region=-1280,1280,1 --y-region=-1280,1280,1 --overall-angle -360 --verbose --slices-per-device 125 --axis-angle-y 0 --axis-angle-x 32.5 --output /data/visitor/sc4593/id19/Lukas_8bit/./set15_SCW3_zoneL_mid_325nm_35keV_real_40mm/xySlices --fix-nan-and-inf --z-parameter z --region=-900.000,900.000,14.400 --center-position-x 1316.25 --center-position-z 1080.5 --volume-angle-x 0 --volume-angle-y 0 --gpus 1 --num-gpu-threads 1
> DEBUG: UFO duration: 698.37 s (too slow!)
> [similar duration times obtained using two GPUs]
> # now restricting to device 1:
> 
> CUDA_VISIBLE_DEVICES=1
> tofu reco --burst 8 --retry-timeout 10 --retries 6 --number 4200 --step 1 --darks /run/shm/sc4593//set15_SCW3_zoneL_mid_325nm_35keV_real_40mm/darkend0000.edf --dark-scale .02000 --flats /run/shm/sc4593//set15_SCW3_zoneL_mid_325nm_35keV_real_40mm/ref*_0000.edf --flats2 /run/shm/sc4593/set15_SCW3_zoneL_mid_325nm_35keV_real_40mm/ref*_4200.edf --projections /run/shm/sc4593/set15_SCW3_zoneL_mid_325nm_35keV_real_40mm/set15_SCW3_zoneL_mid_325nm_35keV_real_40mm*.edf --absorptivity --pixel-size .000000325 --projection-padding-mode clamp_to_edge --genreco-padding-mode clamp_to_edge --x-region=-1280,1280,1 --y-region=-1280,1280,1 --overall-angle -360 --verbose --slices-per-device 125 --axis-angle-y 0 --axis-angle-x 32.5 --output /data/visitor/sc4593/id19/Lukas_8bit/./set15_SCW3_zoneL_mid_325nm_35keV_real_40mm/xySlices --fix-nan-and-inf --z-parameter z --region=-900.000,900.000,14.400 --center-position-x 1316.25 --center-position-z 1080.5 --volume-angle-x 0 --volume-angle-y 0 --gpus 0 --num-gpu-threads 1
> DEBUG: UFO duration: 376.32 s (OK!)

@xeiwoka I am very sorry to reply this late. I just tested the commands above and there you set number of slices to be exactly the number of total slices, i.e. only one GPU will do the reconstruction, whereas the rest of the processing steps (like the flat correction) takes place on both GPUs, i.e. the data needs to be copied from one to the other, which takes a lot of time.

**Comment by milkdrop** [comment]
Hi Tomas, 

first I decided that my comment are entirely unrelated to the original issue, but now I'd like to clarify the situation. You mentioned that "whereas the rest of the processing steps (like the flat correction) takes place on both GPUs". In my tests the only case when server with two GPUs was faster is when algorithm is able to use all GPU RAM (i.e. tofu reco). IN all other cases including ufo-filters and tofu tomo it made no difference at all that one server has twice more TFLOPs and four times more GPU memory. So can ufo-filters make proper use of multiple GPU devices? 

Cheers,
Sergei.

**Comment by tfarago** [comment]
That is a bit strange, of course GPUs should be kept busy to hide all the data copying but sinogram-based reconstruction must also benefit from more cards. Please run
```bash
ufo-launch dummy-data width=2048 height=3000 number=1000 ! fft ! filter ! ifft ! backproject ! null
```
with two cards and then also (on the same machine) specify `export UFO_DEVICES=0`. You should see the result. I see 2x speedup with 2 cards.

**Comment by milkdrop** [comment]
Hi, 

====First question - benchmarking with dummy data ======
I do see expected speedup as you predicted:
ufo-launch dummy-data width=2048 height=3000 number=1000 ! fft ! filter ! ifft ! backproject ! null
1000 items processed ...
Finished in 29.50144s
-bash-4.2$ export UFO_DEVICES=0
-bash-4.2$ ufo-launch dummy-data width=2048 height=3000 number=1000 ! fft ! filter ! ifft ! backproject ! null
1000 items processed ...
Finished in 55.21167s
Moreover on PC with one RTX4000 it is also considerably slower:
ufo-launch dummy-data width=2048 height=3000 number=1000 ! fft ! filter ! ifft ! backproject ! null
1000 items processed ...
Finished in 88.61012s

...so why it is not the case when I work with images? Does setup of HDDs limit the overall perfromance?

====Second question - reconstruction from sinograms =======
2xP5000 (hardware RAID0)
time tofu tomo --sinograms tt-000-sinos --output tt-tomo-sli/sli-%04i.tif --axis 1014
INFO: Execution time: 40.311375 s
export UFO_DEVICES=0
-bash-4.2$ time tofu tomo --sinograms tt-000-sinos --output tt-tomo-sli/sli-%04i.tif --axis 1014    INFO: Execution time: 39.825661 s
1xRTX4000 (single ultra-fast SSD)
 time tofu tomo --sinograms tt-000-sinos --output tt-tomo-sli/sli-%04i.tif --axis 1014
INFO: Execution time: 33.719641 s

**Comment by milkdrop** [comment]
OK, careful benchmarking shows that the write speed is the problem. Thanks for your help!

**Comment by tfarago** [comment]
> tofu reco (laminographic axis tilt 32.5°) considerably slows down (~2x duration) when using flat-field correction (ffc) on multiple GPUs. Without ffc and using --only-bp the durations increase by only some tens of percent.
> 
> This also happens when one uses the device 1 out of 0,1.
> 
> Examples on 2x Tesla K20m:
> # Using device 1 out of 0,1:
> 
> #DEBUG Duration: --gpus 1 --num-gpu-threads 1
> tofu reco --burst 8 --retry-timeout 10 --retries 6 --number 4200 --step 1 --darks /run/shm/sc4593//set15_SCW3_zoneL_mid_325nm_35keV_real_40mm/darkend0000.edf --dark-scale .02000 --flats /run/shm/sc4593//set15_SCW3_zoneL_mid_325nm_35keV_real_40mm/ref*_0000.edf --flats2 /run/shm/sc4593/set15_SCW3_zoneL_mid_325nm_35keV_real_40mm/ref*_4200.edf --projections /run/shm/sc4593/set15_SCW3_zoneL_mid_325nm_35keV_real_40mm/set15_SCW3_zoneL_mid_325nm_35keV_real_40mm*.edf --absorptivity --pixel-size .000000325 --projection-padding-mode clamp_to_edge --genreco-padding-mode clamp_to_edge --x-region=-1280,1280,1 --y-region=-1280,1280,1 --overall-angle -360 --verbose --slices-per-device 125 --axis-angle-y 0 --axis-angle-x 32.5 --output /data/visitor/sc4593/id19/Lukas_8bit/./set15_SCW3_zoneL_mid_325nm_35keV_real_40mm/xySlices --fix-nan-and-inf --z-parameter z --region=-900.000,900.000,14.400 --center-position-x 1316.25 --center-position-z 1080.5 --volume-angle-x 0 --volume-angle-y 0 --gpus 1 --num-gpu-threads 1
> DEBUG: UFO duration: 698.37 s (too slow!)
> [similar duration times obtained using two GPUs]
> # now restricting to device 1:
> 
> CUDA_VISIBLE_DEVICES=1
> tofu reco --burst 8 --retry-timeout 10 --retries 6 --number 4200 --step 1 --darks /run/shm/sc4593//set15_SCW3_zoneL_mid_325nm_35keV_real_40mm/darkend0000.edf --dark-scale .02000 --flats /run/shm/sc4593//set15_SCW3_zoneL_mid_325nm_35keV_real_40mm/ref*_0000.edf --flats2 /run/shm/sc4593/set15_SCW3_zoneL_mid_325nm_35keV_real_40mm/ref*_4200.edf --projections /run/shm/sc4593/set15_SCW3_zoneL_mid_325nm_35keV_real_40mm/set15_SCW3_zoneL_mid_325nm_35keV_real_40mm*.edf --absorptivity --pixel-size .000000325 --projection-padding-mode clamp_to_edge --genreco-padding-mode clamp_to_edge --x-region=-1280,1280,1 --y-region=-1280,1280,1 --overall-angle -360 --verbose --slices-per-device 125 --axis-angle-y 0 --axis-angle-x 32.5 --output /data/visitor/sc4593/id19/Lukas_8bit/./set15_SCW3_zoneL_mid_325nm_35keV_real_40mm/xySlices --fix-nan-and-inf --z-parameter z --region=-900.000,900.000,14.400 --center-position-x 1316.25 --center-position-z 1080.5 --volume-angle-x 0 --volume-angle-y 0 --gpus 0 --num-gpu-threads 1
> DEBUG: UFO duration: 376.32 s (OK!)

So here are the latest numbers (the table is not reproduced nicely and thus hard to read, sorry, columns are the number of used GPUs, last one means I used CUDA_VISIBLE_DEVICES=0, rows are then different tofu reco commands):
UFO Setting \ GPU Setting                   1           2           1 visible to UFO

   ffc 125 slices per gpu, 125 total:        32.80       32.72       32.49
no-ffc 125 slices per gpu, 125 total:      30.22       30.54       30.07
   ffc 400 slices per gpu, 400 total:        52.98       51.16       51.56
   ffc 400 slices per gpu, 800 total:        101.18      56.12       104.97

The first column is the same setting as in the original comment: 125 slices per GPU and 125 slices to reconstruct in total, there should be no performance difference because everything fits to one card. Second row shows the same without doing flat correction, third just increases the amount of slices and most importantly, the last row has double the amount of slices as the number of slices per one card and there I see the cca. 2x speedup.

To conclude, there must have been something weird going on when you did the benchmark. Could the system have been under different loads at the times of the two commands? What would be nice is to see the output of the `tofu reco --verbose` commands to make sure the work was not split between the two cards. If you can re-do them that would be nice, or at some point in the future when you encounter the same issue please keep this in mind and pot it then.

If you agree, I would close the issue for now.
